Improve TEP for Env in Pod Template (#708)
tektoncd/community: teps/0101-env-in-pod-template.md

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> Providing the users a functionality of specifying additional environment variables or overwriting existing environment 

Reviewer: ```suggestion
Providing the users a functionality for specifying additional environment variables or overwriting existing environment 
```

Author: corrected

variables defined in `Steps` and `StepTemplate` of `Task` at runtime through `TaskRun` and `PipelineRun`. In order to 
achieve that, a [Pod template](https://github.com/tektoncd/pipeline/blob/main/docs/podtemplates.md) should support 
configuration of environment variables, which are combined with those defined in `steps` and `stepTemplate`, and then 

> passed to all step containers. That allows to exclude common variables to the global level as well as overwrite 

Reviewer: ```suggestion
passed to all step containers. This let the user excluding common variables to the global level as well as overwrite 
```

Author: corrected

specified in a `stepTemplate` in the single place when running pipelines.
That helps to optionally pass settings, without a need to define additional pipeline parameters.

Also, users can specify the global environment variables which needs to be added for every `TaskRun` or

> `PipelineRun`, then they not need to specify that every time in all `Tasks` and changing the values of that will also

Reviewer: i think it may need some rephrasing, i am not sure if i understand the meaning without someone explaining it to me :)

Author: corrected

This might be a good place to talk about core concepts and how they relate.
-->

1. Forbidden Environment Variables are for pod template of `TaskRun` and `PipelineRun`, are they also valid for global 
   default pod Template? How will it be avoided then? How user will get the error for that?

2. If forbidden environment variables are specified, should they be avoided at Task's `stepTemplate` field or Step's 
   `Env` field too ?

> 3. Will there be option to pass different values of a same env, for different task in a pipelines? There can be case that

Reviewer 2: ```suggestion
3. Will there be option to pass different values of the same env, for different task in a pipeline? There can be a case that
```

Author: corrected

>    same env requires different value in different tasks of a pipeline, and if we put it at pipelineRun spec level, then we are 

Reviewer 2: ```suggestion
   the same env requires different value in different tasks of a pipeline, and if we put it at pipelineRun spec level, then we are 
```

Author: corrected

>    overwriting in all task.

Reviewer 2: ```suggestion
   overwriting in all tasks.
```

Author: corrected

and merge them till list stage which is converting them to container. The order of Precedence will look like this with 
first being the higher.

    a. Environment Variables in PipelineRun Pod Template or TaskRun Pod Template (Same priority as both are different places)
    c. Global Level Default Environment Variables in Tekton Default Pod Template
    d. Task Level Environment Variables in Task Step Template
    e. Step Level Environment Variables in Step Container Spec

If some default is configured and user is providing as the time of `PipelineRun` and `TaskRun`, then it should overwrite 
the default provided at global level.

#### Forbidden Environment Variables

There should be way to define the forbidden environment variables, which should not be overwritten by `PipelineRun` or 
`TaskRun`. If done so, webhook will throw an error that these are forbidden envs.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-defaults
  namespace: tekton-pipelines
  labels:
    app.kubernetes.io/instance: default
    app.kubernetes.io/part-of: tekton-pipelines
data:

>   pod-template-rules: | {

Reviewer: does it need to be a json inside yaml?

Author: yay, this needs to be improved, I will update it. we should have only yaml

Author: fixed
